File: democpp/surface1.cpp
#include "lmfit.hpp"
#include <iostream>

double f(double tx, double tz, const double *p)
{
    return p[0] + p[1]*tx + p[2]*tz;
}

typedef struct {
    const double *tx;
    const double *tz;
    const double *y;
    double (*const f)(double tx, double tz, const double *p);
} data_struct;

void evaluate_surface(const double *par, int m_dat, const void *data,
                      double *fvec, int *info)
{
    data_struct *D = (data_struct*)data;

    for (int i = 0; i < m_dat; i++)
        fvec[i] = D->y[i] - D->f(D->tx[i], D->tz[i], par);
}

int main()
{
    std::vector<double> par0{ -1, 0, 1 };

    const std::vector<double> tx{ -1, -1, 1, 1};
    const std::vector<double> tz{ -1, 1, -1, 1};
    const std::vector<double> y{ 0, 1, 1, 2};
    const int m_dat = y.size();
    assert((int) tx.size() == m_dat);
    assert((int) tz.size() == m_dat);

    const data_struct data = {tx.data(), tz.data(), y.data(), f};

    lm_control_struct control = lm_control_double;
    control.verbosity = 9;

    std::cout << "Fitting:" << '\n';
    auto result = lmfit::minimize(par0, (const void*) &data, m_dat, &evaluate_surface, &control);

    std::cout << "\nResults:" << '\n';
    std::cout << "status after " << result.status.nfev
              << " function evaluations:\n" << lm_infmsg[result.status.outcome]
              << '\n';
    std::cout << "obtained parameters:" << '\n';
    for (size_t j = 0; j < result.par.size(); ++j)
        std::cout << "par[" << j << "] = " << result.par[j] << '\n';

    if (result.status.fnorm > 1e-14) {
        std::cout << "FAILURE (obtained norm = " << result.status.fnorm << " is too large)\n";
        return 1;
    }
    std::cout << "SUCCESS (obtained norm = " << result.status.fnorm << ")\n";
    return 0;
}

File: pythondemo/surface1.py
import swig.pylmfit as pylm

class Data_t(object):
    y: list()
    tx: list()
    tz: list()
    def __init__(self, y, tx, tz):
        self.y = y
        self.tx = tx
        self.tz = tz

    def f(self, tx, tz, p):
        return p[0] + p[1]*tx + p[2]*tz

def eval_surface(p, data=None):
    fvec = []
    for i in range(4):
        fvec.append(data.y[i] - data.f(data.tx[i], data.tz[i], p))
    return (fvec, 0)

def surface1():
    par0 = pylm.VectorDouble([-1, 0, 1])
    tx = [-1, -1, 1, 1]
    tz = [-1, 1, -1, 1]
    y = [0, 1, 1, 2]
    data = Data_t(y, tx, tz)
    wrap = pylm.Data_Wrap(data)
    pylm.set_parlen(3)
    pylm.set_callback(eval_surface)
    control = pylm.lm_control_double

    print("Fitting:")
    result = pylm.minimize(par0, pylm.to_void(wrap), 4, pylm.function_eval, control)

    print("\nResults:")
    print(f"status after {result.status.nfev} function evaluations:")
    pylm.print_infmsg(result.status.outcome)
    print("Obtained parameters:")
    for j in range(par0.size()):
        print(f"par[{j}] = {result.par[j]}")

    print("Fitting data as follows:")
    for i in range(len(tx)):
        ff = data.f(tx[i], tz[i], result.par)
        print(f"  t[{i}]={tx[i]},{tz[i]}  y={y[i]} fit={ff} residue={y[i]-ff}")

    if (result.status.fnorm > 1e-14):
        print(f"FAILURE (obtained norm = {result.status.fnorm} is too large)")
        return 1
    print(f"SUCCESS (obtained norm = {result.status.fnorm})")
    return 0

if __name__ == "__main__":
    print(surface1())
